### PR TITLE
add typing_extensions to dependencies

### LINKED ISSUE #25974: Add dependency typing_extensions for type hints
[Typing Extensions](https://pypi.org/project/typing-extensions/) – Backported and Experimental Type Hints for Python.

To make use of Type Hints only available in python 3.7 and some other Experimental Type Hints (like `Final`) this package is required. Discussion in #25961

### PR DESCRIPTION
**Pull request by vaibhavhrt**
- [ ] closes #25974 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by vaibhavhrt** [comment]
@WillAyd I have added `typing_extensions`(without freezing the version) to both `environment.yml` and `requirements-dev.txt` manually because I am not using anaconda(not a big fan of anaconda). Take a look if everything looks alright.

**Comment by WillAyd** [comment]
Can you check master in a 3.5.1 environment? Might need to import Type from this library as part of this

**Comment by vaibhavhrt** [comment]
@WillAyd  I am not sure what you mean.
Should I import `Type` from `typing_extensions` and then run tests in python 3.5.1 locally?

**Comment by WillAyd** [comment]
Can you check if current master runs on 3.5.1 first? If not then yes might need to import from type_extensions to resolve

**Comment by jreback** [comment]
we should pin one of the builds to 3.5.1

separately not averse to bumping to 3.5.3 (or whatever we need)

File: environment.yml [hunk 1]
```diff
@@ -25,6 +25,7 @@ dependencies:
   - pytest-mock
   - sphinx
   - numpydoc
+  - typing_extensions
```
Comment (WillAyd) [inline environment.yml]: We might need to require this if it causes import failures with 3.5.1

**Comment by WillAyd** [comment]
Makes sense. I think we could pin 3.5.1 as part of this PR and tackle maybe bumping to 3.5.3 separately if we wanted to

File: environment.yml (hunk 1 above)
Comment (vaibhavhrt) [reply environment.yml]: I am setting up the environment and C compiler on my laptop. I will run tests and let you know if it fails.

**Review by jreback — CHANGES_REQUESTED** [review]
you need to add this to all of the ci builds in the pandas/ci/deps dir

**Review by jreback — CHANGES_REQUESTED** [review]
need to update setup.py & the install docs as well.

File: environment.yml (hunk 1 above)
Comment (vaibhavhrt) [reply environment.yml]: @WillAyd does it matter how I create virtualenv(anaconda, pipenv) when running tests locally. My system is low on space and anaconda takes up a lot of space. So is it ok if I use pipenv to create env and then run `pytest pandas` or will that make any difference.

**Comment by WillAyd** [comment]
We provide instructions for conda but you could get to the same place using pip or pipenv for sure

**Comment by vaibhavhrt** [comment]
@WillAyd I finally managed to get MS Build Tools 14 working and build pandas. I should be able to run tests locally now. But I am too busy at work today. Tomorrow I will run tests on python 3.5.1 and let you know the results. I will take care of changes requested by @jreback too tomorrow.

**Comment by TomAugspurger** [comment]
What exactly are we using from this? If we’re only using one or two definitions, I’d rather just include those in our typing compat module. Adding a new dependency shouldn’t be done lightly.

**Comment by WillAyd** [comment]
@TomAugspurger what's the concern with third party library? There's a lot of active development with types so I think it would be more work to vendor pieces of it and keep in sync rather than to include as dependency

**Comment by vaibhavhrt** [comment]
@WillAyd on python 3.5.1 even master will fail. I can't even import pandas from local build. Here is relevant part of traceback:

    ...
      File "C:\Users\Harish\Documents\GitHub\pandas\pandas\core\dtypes\base.py", line 2, in <module>
      from typing import List, Optional, Type
    ImportError: cannot import name 'Type'

[typing.Type](https://docs.python.org/3/library/typing.html#typing.Type) was added in 3.5.2

**Comment by vaibhavhrt** [comment]
After adding typing_extensions to dependency and importing `Type` from `typing_extensions` 
I got the following test results:

     22 failed, 47816 passed, 1368 skipped, 1028 xfailed, 15 xpassed, 3469 warnings in 4985.95 seconds

I can't see how any of the changes I made can make tests fail. I have pushed the changes, you can see tests results on travis and let me know how to fix it.

**Comment by vaibhavhrt** [comment]
nvm I see the problem. My bad

**Comment by vaibhavhrt** [comment]
@WillAyd tests are passing on travis but not locally, don't know why. Also some azure tests for mac and linux are failing. Do I remove `typing-extensions` from yml of those tests or should I install them from pip instead of conda?

**Comment by WillAyd** [comment]
You’d need to be explicit about which failures you are getting locally

**Comment by vaibhavhrt** [comment]
@WillAyd running tests locally takes a long time. I will worry about that some other time. Travis tests(for my fork too) is passing, so there is something wrong with my system. For now please let me know to fix the Azure tests(some of them are failing). They all have similar issue:

    conda env create -q --file=ci/deps/azure-macos-35.yaml
    Warning: you have pip-installed dependencies in your environment file, but you do not list pip itself as one of your conda dependencies.  Conda may not use the correct pip to install your packages, and they may end up in the wrong place.  Please add an explicit pip dependency.  I'm adding one for you, but still nagging you.
    Collecting package metadata: ...working... done
    Solving environment: ...working... failed
    WARNING: The conda.compat module is deprecated and will be removed in a future release.

    ResolvePackageNotFound: 
      - typing-extensions

    ##[error]Bash exited with code '1'.

I think I need to change to something in the the `yaml` files of these tests but not sure what.

**Comment by TomAugspurger** [comment]
conda-forge doesn't build packages for 3.5 (which demonstrates one issue
with taking on dependencies @WillAyd).

**Comment by WillAyd** [comment]
Hmm ok. So you think better to just add Type and Final to our internal typing package? Not opposed if that’s easier

**Comment by jreback** [comment]
> Hmm ok. So you think better to just add Type and Final to our internal typing package? Not opposed if that’s easier

yes let's just add what we need to _typing, rather that this dep.

**Comment by vaibhavhrt** [comment]
@jreback if no one is already working on adding `Type` and `Final` to `pandas.typing` I can do that. Let me know if I should open a new issue and PR for that. I am closing this PR.

**Pull request closed without merging**